### ISSUE TITLE
template in template paradigm cripples dev experience

### ISSUE THREAD

**Issue by Morriz** [codecentric/helm-charts#343]
Hi guys, thanks first of all for taking on the creation and maintenance of these charts.

Now, I would ask you to not stray from chart best practices (and software development principles for that matter), and don't resort to such mechanisms as in this example:
```yaml
          {{- with .Values.keycloak.extraVolumeMounts}}
            {{- tpl . $ | nindent 12 }}
          {{- end }}
```
Why? (Please bear with my style of reasoning as I go along ;)

**Facts**:
1.`extraVolumeMounts` is a string that is used as a template, with the $ context (which is the root context object of the helm chart) given to the template
2. this effectively nukes the option to pass in yaml as is (which is recognizable inside a yaml file by good editors) and thus also nukes the powers that come with yaml detection (autocompletion, linting etc)
3. This technique of offering input handles is used consistently throughout the chart

**Assumption**:
It looks as if the chart creator anticipates a perceived need: this particular technique of bringing values into the chart, allowing users to provide a template that the chart will populate with `$`

Now, if the creator is aware of the costly tradeoff I described in nr 2 I don't know, but if my assumption is correct, then I would like to postulate these

**Counter arguments**:
- there exist other ways of providing values to the chart: plain yaml. Let the user choose a solution to craft it, be it with kustomize, helmfile, whatever
- it is going against seperation of concerns: it not only forces how users should provide input, (crippling development workflow with their input yaml), but it also forces them to gain awareness of the new options given to them (what exists on the $ object?).

In fact, I would like to make a bold statement (that I could see proven wrong):

**No real functionality is gained.**

Why? Most of what `$` has to offer is already available to the user: .Values (the user is the one bringing together all the input in the first place). All the other stuff really seems to me not interesting to the user:
- `.Template`: what could you possibly want to use it for when providing input?
- `.Capabilities`: same
- `.Release`, `.Files`, `.Chart`, etc: again, what would you need it for when providing input?

Maybe there is room to anticipate weird edge cases here, but you get my drift. If there are real world use cases I am missing here, then please let us know.

**Conclusion**:
All in all this seems to me as a case of YAGNI, overengineering and possibly a lack of knowledge of software design. But, to be fair, I would like to ask some questions to the creators first:
- Can you present any use cases that demonstrate the need for this? Are you already relying on this technique and is that documented somewhere?
- What is your rationale when explaining any anticipated advantages over these real world losses?.

Fnally, I would propose to NOT obfuscate yaml (or any structured text for that matter) in favor of a regular string, and just expect the yaml to come in as is. So I would just rewrite the above snippet as:
```yaml
          {{- with .Values.keycloak.extraVolumeMounts }}
            {{- toYaml . | nindent 12 }}
          {{- end }}
```

And follow that reasoning consistently when creating charts.

TLDR

KISS, but I sincerely hope you took the time to read what I wrote ;p

**Comment by Morriz** [comment]
Ok, I have seen one example in this repo (but I don't agree with its reasoning): #43. That goes against another important concern imo: mixing of alien resources (the configmap mentioned) with resources under management of the chart.
 
Anyway, imo most additions to the chart yaml should be in the form of a PR if it serves all of us. I guess it all comes down to looking precisely in each occasion what makes most sense to do: trade in predictability for flexibility when it is really needed. But is that really ever needed? I intuitively feel that there is always a better solution than to offer templates to "solve" problems that imo are the result of misuse/misunderstanding.

**Comment by Morriz** [comment]
Having inspected the whole chart I come to see that `tpl` brings joy in some areas like annotations (which brings no drawback there as the values are already string). However, in this list are imo only 3 valid candidates for tpl usage:

* `keycloak.extraInitContainers`
* `keycloak.extraContainers`
* `keycloak.extraEnv`
* `keycloak.affinity` <- contrived example uses it
* `keycloak.extraVolumeMounts`
* `keycloak.extraVolumes`
* `keycloak.livenessProbe` <- allowing to use path configuration
* `keycloak.readinessProbe` <- same

The rest of these cases I can't envision a use case for. 

I would really like it if the usage of tpl could be limited to the areas where templating is expected to be useful.

**Comment by unguiculus** [comment]
Thanks for putting this together. I guess there's always two sides of the medal. Look at the readme. There are several examples that make use of this feature. This is a fairly complex chart which warrants using `tpl` in various places. I probably wouldn't use it for simple charts, but here it makes sense IMO. There are no chart best practices that tell you not to use `tpl`. Otherwise, it wouldn't even have been added to Helm in the first place. Templating of values has actually been an oft-requested feature. `tpl` is the solution for this. Also, your referring to other tools shows that something like this is indeed necessary. I'm a Helm maintainer myself and actually helped shape the best practices we applied in the community charts repo. That being said, I'm not inclined to change this.

**Comment by Morriz** [comment]
I guess it is a question of maintaining chart quality. What are the principles, and in what order are they valued?

IMO at least one of them should be typed values, which can be enforced/supported by publishing a values jsonschema. This aids code completion and validation in editors. Templating kills this feature. And what do we get in return? Obscurity?
